packages/hop-node/src/vault/YearnVault.ts

```diff
@@ -0,0 +1,226 @@
+    const url = getRpcUrl(chain)
+    const provider = new Provider(url!, signer)
+    this.signer = signer
+    this.decimals = getTokenDecimals(token)
+    if (!instanceCache[chain]) {
+      instanceCache[chain] = new Yearn(chainId, {
+        provider: {
+          write: provider,
+          read: provider
+        }
+      })
+    }
+    this.yearn = instanceCache[chain]
+  }
+
+  async getBalance (account?: string): Promise<BigNumber> {
+    if (!account) {
+      account = await this.signer.getAddress()
+    }
+    const { vault } = addresses[this.token]
+    const contract = this.getErc20(vault)
+    return contract.balanceOf(account)
+  }
+
+  async deposit (amount: BigNumber) {
+    const account = await this.signer.getAddress()
+    const { token, vault } = addresses[this.token]
+    const needsApproval = await this.needsDepositApproval(amount)
+    if (needsApproval) {
+      console.log('needs approval; sending approval tx')
+      const tx = await this.approveDeposit()
+      console.log('approval tx:', tx.hash)
+      await tx.wait()
+    }
+    console.log('attempting to deposit')
+    const tx = await this.yearn.vaults.deposit(vault, token, amount.toString(), account, {
+      slippage: this.slippage
+    })
+    return tx
+  }
+
+  async withdraw (amount: BigNumber) {
+    const account = await this.signer.getAddress()
+    const { token, vault } = addresses[this.token]
+    const needsApproval = await this.needsWithdrawalApproval(amount)
+    if (needsApproval) {
+      console.log('needs approval; sending approval tx')
+      const tx = await this.approveWithdrawal()
+      console.log('approval tx:', tx.hash)
+      await tx.wait()
+    }
+    console.log('attempting to withdraw')
+    const tx = await this.yearn.vaults.withdraw(vault, token, amount.toString(), account, {
+      slippage: this.slippage
+    })
+    return tx
+  }
+
+  async approveDeposit (amount: BigNumber = constants.MaxUint256) {
+    const { token, vault } = addresses[this.token]
+    const contract = this.getErc20(token)
+    return contract.approve(vault, amount)
+  }
+
+  async approveWithdrawal (amount: BigNumber = constants.MaxUint256) {
+    const { vault, zapOut } = addresses[this.token]
+    const contract = this.getErc20(vault)
+    return contract.approve(zapOut, amount)
+  }
+
+  async needsDepositApproval (amount: BigNumber = constants.MaxUint256) {
+    const account = await this.signer.getAddress()
+    const { token, vault } = addresses[this.token]
+    if (token === EthAddress) {
+      return false
+    }
+    const contract = this.getErc20(token)
+    const allowance = await contract.allowance(account, vault)
+    const needsApproval = allowance.lt(amount)
+    return needsApproval
+  }
+
+  async needsWithdrawalApproval (amount: BigNumber = constants.MaxUint256) {
+    const account = await this.signer.getAddress()
+    const { token, vault, zapOut } = addresses[this.token]
+    if (token !== EthAddress) {
```

Comment:
It's needed to withdraw WETH since that's the token the vault uses for ETH

You can see the yearn UI has an approval for ETH withdrawal

![DeepinScreenshot_select-area_20220307182107](https://user-images.githubusercontent.com/168240/157153490-b7c46a67-8dee-4c27-a56b-6e9df6e3f290.png)

Reply:
Just played with yearn and see the same, so I believe you are correct. I guess I'm just confused about the difference between the ETH vault (which needs approval) vs. the token vaults (which don't need approval). Anyway, LGTM 👍